config/src/main/java/org/springframework/security/config/annotation/web/configuration/WebSecurityConfiguration.java

```diff
@@ -95,12 +100,27 @@ public static DelegatingApplicationListener delegatingApplicationListener() {
 	public Filter springSecurityFilterChain() throws Exception {
 		boolean hasConfigurers = webSecurityConfigurers != null
 				&& !webSecurityConfigurers.isEmpty();
-		if (!hasConfigurers) {
+		boolean hasFilterChain = !securityFilterChains.isEmpty();
+		if (hasConfigurers && hasFilterChain) {
+			throw new IllegalStateException(
+					"Found WebSecurityConfigurerAdapter as well as SecurityFilterChain." +
```

Comment:
Is there a reason to limit to one or the other?

Reply:
If we allow both, it brings up concerns around ordering.

Using the `@Order` annotation doesn't make sense because they are different components, so their order cannot be compared.

The other option I see is to always apply one before the other, for example, first apply all `WebSecurityConfigurerAdapter` configurations and then all `SecurityFilterChain` configurations. But I can see that causing more confusion.